pydantic_ai_slim/pydantic_ai/embeddings/cohere.py

```diff
@@ -0,0 +1,198 @@
+from collections.abc import Sequence
+from dataclasses import dataclass, field
+from typing import Any, Literal, cast
+
+from pydantic_ai.exceptions import ModelAPIError, ModelHTTPError, UnexpectedModelBehavior
+from pydantic_ai.providers import Provider, infer_provider
+from pydantic_ai.usage import RequestUsage
+
+from .base import EmbeddingModel, EmbedInputType
+from .result import EmbeddingResult
+from .settings import EmbeddingSettings
+
+try:
+    from cohere import AsyncClientV2
+    from cohere.core.api_error import ApiError
+    from cohere.core.request_options import RequestOptions
+    from cohere.types.embed_by_type_response import EmbedByTypeResponse
+    from cohere.types.embed_input_type import EmbedInputType as CohereEmbedInputType
+    from cohere.v2.types.v2embed_request_truncate import V2EmbedRequestTruncate
+
+    from pydantic_ai.providers.cohere import CohereProvider
+except ImportError as _import_error:
+    raise ImportError(
+        'Please install `cohere` to use the Cohere embeddings model, '
+        'you can use the `cohere` optional group — `pip install "pydantic-ai-slim[cohere]"`'
+    ) from _import_error
+
+LatestCohereEmbeddingModelNames = Literal[
+    'embed-v4.0',
+    'embed-english-v3.0',
+    'embed-english-light-v3.0',
+    'embed-multilingual-v3.0',
+    'embed-multilingual-light-v3.0',
+]
+"""Latest Cohere embeddings models."""
+
+CohereEmbeddingModelName = str | LatestCohereEmbeddingModelNames
+"""Possible Cohere embeddings model names."""
+
+# Taken from https://docs.cohere.com/docs/cohere-embed
+_MAX_INPUT_TOKENS: dict[CohereEmbeddingModelName, int] = {
+    'embed-v4.0': 128000,
+    'embed-english-v3.0': 512,
+    'embed-english-light-v3.0': 512,
+    'embed-multilingual-v3.0': 512,
+    'embed-multilingual-light-v3.0': 512,
+}
+
+
+class CohereEmbeddingSettings(EmbeddingSettings, total=False):
+    """Settings used for a Cohere embedding model request."""
+
+    # ALL FIELDS MUST BE `cohere_` PREFIXED SO YOU CAN MERGE THEM WITH OTHER MODELS.
```

Comment:
if this is the case, should we just make `cohere` a top level field? Are there any fields that _are_ shared between providers? I guess this isn't that important